cmd/duffle/upgrade.go

```diff
@@ -24,78 +25,80 @@ If no parameters are passed, the parameters from the previous release will be us
 are specified, the parameters there will be used (even if the resolved set is empty).
 `
 
-var upgradeDriver string
-
 type upgradeCmd struct {
-	duffleCmd
-	name       string
-	valuesFile string
-	setParams  []string
-	insecure   bool
-	setFiles   []string
+	out              io.Writer
+	name             string
+	driver           string
+	valuesFile       string
+	bundle           string
+	bundleFile       string
+	setParams        []string
+	insecure         bool
+	setFiles         []string
+	credentialsFiles []string
 }
 
-func newUpgradeCmd() *cobra.Command {
-	uc := &upgradeCmd{}
-
-	var (
-		credentialsFiles []string
-		bundleFile       string
-	)
+func newUpgradeCmd(w io.Writer) *cobra.Command {
+	upgrade := &upgradeCmd{out: w}
 
 	cmd := &cobra.Command{
 		Use:     "upgrade NAME [BUNDLE]",
 		Short:   upgradeUsage,
 		Long:    upgradeLong,
-		PreRunE: uc.Prepare(),
+		Args:    cobra.ExactArgs(1),
+		PreRunE: upgrade.setup(),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			if len(args) != 1 {
-				return errors.New("This command requires at least 1 argument: the name of the installation to upgrade")
-			}
-			uc.name = args[0]
-			uc.Out = cmd.OutOrStdout()
-			bundleFile, err := optBundleFileOrArg2(args, bundleFile, uc.Out, uc.insecure)
-			if err != nil {
-				return err
-			}
-
-			return uc.upgrade(credentialsFiles, bundleFile)
+			upgrade.name = args[0]
+
+			return upgrade.run()
 		},
 	}
 
 	flags := cmd.Flags()
-	flags.StringVarP(&bundleFile, "file", "f", "", "Set the bundle file to use for upgrading")
-	flags.StringVarP(&upgradeDriver, "driver", "d", "docker", "Specify a driver name")
-	flags.StringArrayVarP(&credentialsFiles, "credentials", "c", []string{}, "Specify credentials to use inside the CNAB bundle. This can be a credentialset name or a path to a file.")
-	flags.StringVarP(&uc.valuesFile, "parameters", "p", "", "Specify file containing parameters. Formats: toml, MORE SOON")
-	flags.StringArrayVarP(&uc.setParams, "set", "s", []string{}, "Set individual parameters as NAME=VALUE pairs")
-	flags.BoolVarP(&uc.insecure, "insecure", "k", false, "Do not verify the bundle (INSECURE)")
-	flags.StringArrayVarP(&uc.setFiles, "set-file", "i", []string{}, "Set individual parameters from file content as NAME=SOURCE-PATH pairs")
+	flags.StringVarP(&upgrade.driver, "driver", "d", "docker", "Specify a driver name")
+	flags.StringVarP(&upgrade.bundle, "bundle", "b", "", "bundle to use for upgrading")
```

Comment:
A similar question here as posed around the `uninstall` action cmd.  

Looking to understand how supplying a bundle is intended to affect an `upgrade` action?  I think I can envision the use case a bit more clearly with upgrade: a user has installed a bundle with a given version of a bundle.json and then wishes to upgrade the installation with a revised version of the bundle.json.

If I'm on the right track, I guess my question becomes: Have we yet explored the intricacies and potentially wide-ranging margin for errors when enabling the upgrade of an installation with a different version of a bundle.json (may possibly be vastly different) than was used for the initial installation?

Reply:
See comment above. I think it addresses the same question. Perhaps we should open up an issue to discuss this further.